### PR TITLE
add a Dockerfile and docker-based dev environment

### PR DESCRIPTION
**Pull request by ceralena**
The core of this commit is:

- the Dockerfile, with a multi-staged build that can provide a dev
  environment and also a relatively clean final image
- `docker-compose.yml`, enabling local development
- some boilerplate in `auto/` and `scripts/`: this is something I've
  ripped off directly from work. Everything in `scripts` runs a command directly; everything in `auto/` wraps it in a container

I've added `foreman` for running the local dev server. I've also started
to fill out the README. The table of contents was generated by the
"Markdown All-In-One" VS Code plugin.

This is all working relatively well for me except that the server is crashing on startup with:

```
15:26:20 web.1  | /home/cera/src/cera/exchange/vendor/bundle/ruby/2.6.0/gems/webpacker-4.2.2/lib/webpacker/configuration.rb:95:in `rescue in load': Webpacker configuration file not found /home/cera/src/cera/exchange/config/webpacker.  yml. Please run rails webpacker:install Error: No such file or directory @ rb_sysopen - /home/cera/src/cera/exchange/config/webpacker.yml
(RuntimeError)
```

@kiesia - any thoughts?

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+#
+# We install libpq-dev in this stage, because it's required everywhere.
+#
+# To see if there's a new image available, run "docker pull
+# ruby:2.6.5-slim-buster" and inspect the image digest.
+#
+# If you update the minor or patch versions, you should also update the .tool-versions file.
+#
+# Also see https://hub.docker.com/_/ruby
+FROM ruby:2.6.5-slim-buster@sha256:266f8f86a1de812612505aa5d5e7adf5c3ff8b06fb997b95cefa18fd8249fcbe \
```
Comment (kiesia) [inline Dockerfile]: Why slim Buster over Alpine? I don't expect that we'll need any obscure dependencies in the docker image?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+FROM ruby-base AS final
+
+WORKDIR /work
+
+COPY --from=with-gems /usr/local/bundle/ /usr/local/bundle/
+COPY . .
+
+RUN bundle check
+
+ENV RACK_ENV production
```
Comment (kiesia) [inline Dockerfile]: Is this the equivalent of `RAILS_ENV`?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+COPY Gemfile Gemfile.lock .tool-versions ./
+RUN gem install bundler -v ${RUBY_BUNDLER_VERSION}
+RUN bundle config set without 'development test' && \
+  bundle config set deployment 'true' && \
+  bundle config set path '/usr/local/bundle' && \
+  bundle --jobs=4 \
+    && find /usr/local/bundle/ \( -name "*.gem" -o -name "*.c" -o -name "*.o" -o -name "mkmf.log" \) -delete
+
+### final
+FROM ruby-base AS final
```
Comment (kiesia) [inline Dockerfile]: Should we be running a system update here? How are you normally keeping images up to date? I'm never sure if I'm too paranoid with this?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+FROM ruby-base AS dev-environment
+
+RUN apt-get update \
+  && apt-get install -y --no-install-recommends build-essential curl git \
+  && rm -rf /var/lib/apt/lists/*
+
+RUN gem install bundler -v ${RUBY_BUNDLER_VERSION}
+
+### with-gems
+FROM dev-environment AS with-gems
```
Comment (kiesia) [inline Dockerfile]: Should this step be a general builder step, and we run `assets:precompile` here too?

**Comment by kiesia** [comment]
```
15:26:20 web.1  | /home/cera/src/cera/exchange/vendor/bundle/ruby/2.6.0/gems/webpacker-4.2.2/lib/webpacker/configuration.rb:95:in `rescue in load': Webpacker configuration file not found /home/cera/src/cera/exchange/config/webpacker.  yml. Please run rails webpacker:install Error: No such file or directory @ rb_sysopen - /home/cera/src/cera/exchange/config/webpacker.yml
(RuntimeError)
```

Let's just run the rake task and go with the default webpack config for now.

**Comment by kiesia** [comment]
Let's drop down to a normal user in dev and prod?

File: Dockerfile (hunk 3 above)
Comment (ceralena) [reply Dockerfile]: I prefer not to do a system update here as it means builds aren't deterministic. Not sure how to have good discipline about keeping the image up-to-date, though. Perhaps a periodic reminder, or something.

File: Dockerfile (hunk 1 above)
Comment (ceralena) [reply Dockerfile]: I've ran into some issues with the SDKs for AWS and GCP not building properly with musl, which is alpine's libc. I like the elegance of Alpine, but tbh as soon as I realised its libc was giving me issues I ran back for Debian.

File: Dockerfile (hunk 2 above)
Comment (ceralena) [reply Dockerfile]: Yep. Rails will obey both - but some Rack middlewares only obey `RACK_ENV`. Happy to just set `RAILS_ENV`, or both, if you prefer.

**Comment by ceralena** [comment]
> Let's drop down to a normal user in dev and prod?

Good call, I'll add that.

File: Dockerfile (hunk 2 above)
Comment (kiesia) [reply Dockerfile]: No that's all good, thanks. :+1:

File: Dockerfile (hunk 3 above)
Comment (kiesia) [reply Dockerfile]: > builds aren't deterministic

Do you fee that's a bigger problem than remembering to update hashes in the Dockerfile?

File: Dockerfile (hunk 1 above)
Comment (kiesia) [reply Dockerfile]: Great ok then Buster is fine by me. :+1:

File: Dockerfile (hunk 4 above)
Comment (ceralena) [reply Dockerfile]: It may seem like overkill, but I'd prefer to add another stage for that - as it could blow out if we need other build steps.

Comment (kiesia) [reply Dockerfile]: Your call!

File: Dockerfile (hunk 3 above)
Comment (ceralena) [reply Dockerfile]: Yeah, on balance I think running an update's probably better. I'll add it.

File: .dockerignore [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Exclude everything by default
```
Comment (kiesia) [inline .dockerignore]: Nice!

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      PORT: 3000
+    working_dir: /work
+    entrypoint: /work/scripts/bundle-exec
+    command: bash
+    volumes:
+      - .:/work
+      - exchange-web-gem-cache:/usr/local/bundle
+      # Don't mount the .env file or the tmp dir when running in Docker
+      - /dev/null:/work/.env
+      - /dev/null:/work/tmp
```
Comment (kiesia) [inline docker-compose.yml]: Should we throw in postgres:11 here? Then the app will be working.

Comment (ceralena) [reply docker-compose.yml]: Will do.

**Comment by kiesia** [comment]
Thanks @ceralena!!

**Comment by ceralena** [comment]
Just rebased to fix conflicts.

**Pull request merged**